### ISSUE TITLE
Using Skippy with Android Projects

### ISSUE THREAD

**Issue by helios66** [skippy-io/skippy#162]
Hello! Amazing work with this library. I am currently trying to get it setup for a multi-module android project (~400 modules). The project is in Kotlin and used both Junit4 and Junit5. 

I was able to apply the plugin to modules in the project, running tests generates the .skippy folder with the config.json, LATEST, predictions.log and test-impact-analysis.json files. 

However, subsequent runs with or without changes don't seem to update the predictions.log file the way the tutorial [says](https://www.skippy.io/tutorials/skippy-gradle-junit5#setting-up-your-environment). 
All tests annotated by @PredictWithSkippy and using the @ClassRule never get updated from the `{testname}, EXECUTE,TEST_IMPACT_ANALYSIS_NOT_FOUND` phase in the predictions.log file. I would expect that running the tests with no changes should at least update the entries to `{testname}, SKIP,NO_CHANGE` and vice-versa.

I would like to know if there's a configuration somewhere I might be missing?

**Comment by fmck3516** [comment]
@helios66 Following up on the above. Is there a way for you to provide a minimal sample project?

**Comment by Link184** [comment]
@fmck3516 Looks like I have the same problem, please check this sample project.

PS: Amazing tool, I am really excited to see it in action on android!

[TestSkippy.zip](https://github.com/user-attachments/files/17681507/TestSkippy.zip)

**Comment by fmck3516** [comment]
@Link184 Thanks for the test case - I will have a look shortly.

**Comment by sergio-sastre** [comment]
> @fmck3516 Looks like I have the same problem, please check this sample project.
> 
> PS: Amazing tool, I am really excited to see it in action on android!
> 
> [TestSkippy.zip](https://github.com/user-attachments/files/17681507/TestSkippy.zip)

Same here! This looks promising! 😱
@fmck3516 
Let us know if you need some help to add support for Android

**Comment by fmck3516** [comment]
I am able to re-produce the issue (thanks to the attached sample project):

<img width="1776" alt="image" src="https://github.com/user-attachments/assets/a4be1c98-6e82-4548-bfa4-9aeb972af19d">

https://github.com/skippy-io/skippy/blob/07dce1731f16704dc4ffad1bd0216a633e7d2d82/skippy-gradle/src/main/java/io/skippy/gradle/ProjectSettings.java#L37

The way Skippy locates class files does not work for the attached project.

I will release a separate `skippy-gradle-android` plugin to fix that.

> Let us know if you need some help to add support for Android

I am working on a fix, but please throw a PR my way if you have an idea how to fix that. I would love to see other people's solution. And I would definitely appreciate help to thoroughly test the upcoming support for Android.

**Comment by Link184** [comment]
@fmck3516 My assumption is that we need to use `AndroidSourceSetContainer` instead `SourceSetContainer`, for that our plugin must depend on AGP, probably that is the reason why you want to make another artifact just for android.

**Comment by fmck3516** [comment]
> @fmck3516 My assumption is that we need to use `AndroidSourceSetContainer` instead `SourceSetContainer

Can you shoot me a link where I can look up `AndroidSourceSetContainer`? I don't find any references.

> for that our plugin must depend on AGP, probably that is the reason why you want to make another artifact just for android.

Yes, my gut feeling: I don't want to cram the Android dependencies into the Java plugin to keep the dependency footprint in check. The build plugins are thin layers around the core implementation - adding another plugin isn't a big deal.

**Comment by fmck3516** [comment]
> I can create a PR with this POC tomorrow

Awesome!

I've created the boilerplate for the android plugin: https://github.com/skippy-io/skippy/tree/main/skippy-gradle-android

Here is code that needs to figure out how to initialize the location of the class files (and some other stuff) for Android-based builds:

https://github.com/skippy-io/skippy/blob/4d0a4a6f76634cf4f239d2b68d716e9167b34ff4/skippy-gradle-android/src/main/java/io/skippy/gradle/android/ProjectSettings.java#L29-L38

**Comment by fmck3516** [comment]
FYI: https://github.com/skippy-io/skippy/issues/133

My current plan how to handle duplicate class names across output folders for Android-based builds

**Comment by Link184** [comment]
right now the library stopped working in android. In `test-impact-analysis.json`, the `test` object is always empty 🤔

**Comment by fmck3516** [comment]
> right now the library stopped working in android. In `test-impact-analysis.json`, the `tests` object is always empty 🤔

I added a comment in https://github.com/skippy-io/skippy/issues/185

**Comment by fmck3516** [comment]
> right now the library stopped working in android. In `test-impact-analysis.json`, the `tests` object is always empty 🤔

Why i think that this is related to https://github.com/skippy-io/skippy/issues/185.

Due to https://github.com/skippy-io/skippy/issues/185, Skippy does not create the `test-impact-analysis.json` file:

```
git clone git@github.com:skippy-io/skippy-android-sandbox.git
cd skippy-android-sandbox

./gradlew check
BUILD SUCCESSFUL in 10s
```

At this point, `test-impact-analysis.json` should exist, but it doesn't:

```
ls -l app/.skippy

predictions.log
tmp
```

When you run `check` the next time, the test tasks are `UP-TO-DATE`:

```
./gradlew check  --console=plain

Task :app:testReleaseUnitTest UP-TO-DATE
Task :app:testDebugUnitTest UP-TO-DATE
```

This time, `test-impact-analysis.json` is written:

```
ls -l app/.skippy

LATEST
config.json
test-impact-analysis.json
```

But the `tests` array is empty (since no tests were executed):
```
tail app/.skippy/test-impact-analysis.json

    ...
    "tests": [

    ]
}
```

**Comment by fmck3516** [comment]
Skippy support for the provided Android sample project is available as of release 0.0.24.

**Comment by fmck3516** [comment]
Hi all - release `0.0.24` is now available in Gradle's Plugin Portal (https://plugins.gradle.org/plugin/io.skippy.android) and Maven Central.  This release adds support for the Android sample project shared by @Link184. Thanks to everybody for the contributions and discussions. I'm especially pumped that this is the first release that includes an "external" contribution 💪 

I will update and extend the documentation in the next couple of days.

In the meantime: Please take release `0.0.24` (or `0.0.25-SNAPSHOT`) for a spin and let me know how it goes (either in the comments or as new issue).

Full release notes: https://github.com/skippy-io/skippy/releases/tag/v0.0.24

Note that JUnit 4 support is now provided as `@Rule` instead of `@ClassRule`:

```
public class MyTest {

    @Rule
    public TestRule skippyRule = Skippy.predictWithSkippy();

    ...

}
```
